lib/engine-universal/src/artifact.rs

```diff
@@ -1,402 +1,211 @@
-        &self.signatures
+    fn globals(&self) -> &[(GlobalType, GlobalInit)] {
+        &self.local_globals[..]
     }
 
-    fn func_data_registry(&self) -> &FuncDataRegistry {
-        &self.func_data_registry
+    fn start_function(&self) -> Option<FunctionIndex> {
+        self.start_function
     }
-    fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        // Prepend the header.
-        let mut serialized = Self::MAGIC_HEADER.to_vec();
 
-        serialized.resize(SERIALIZED_METADATA_CONTENT_OFFSET, 0);
-        let mut writable_leb = &mut serialized[SERIALIZED_METADATA_LENGTH_OFFSET..];
-        let serialized_data = self.serializable.serialize()?;
-        let length = serialized_data.len();
-        leb128::write::unsigned(&mut writable_leb, length as u64).expect("Should write number");
-
-        let offset = pad_and_extend::<SerializableModule>(&mut serialized, &serialized_data);
-        assert_eq!(offset, SERIALIZED_METADATA_CONTENT_OFFSET);
-
-        Ok(serialized)
+    fn function_by_export_field(&self, name: &str) -> Option<FunctionIndex> {
+        self.exported_functions.get(name).copied()
     }
-}
-
-/// It pads the data with the desired alignment
-pub fn pad_and_extend<T>(prev_data: &mut Vec<u8>, data: &[u8]) -> usize {
-    let align = std::mem::align_of::<T>();
 
-    let mut offset = prev_data.len();
-    if offset & (align - 1) != 0 {
-        offset += align - (offset & (align - 1));
-        prev_data.resize(offset, 0);
+    fn signatures(&self) -> &[wasmer_vm::VMSharedSignatureIndex] {
+        self.signatures.values().as_slice()
     }
-    prev_data.extend(data);
-    offset
-}
-
-#[cfg(test)]
-mod tests {
-    use super::pad_and_extend;
 
-    #[test]
-    fn test_pad_and_extend() {
-        let mut data: Vec<u8> = vec![];
-        let offset = pad_and_extend::<i64>(&mut data, &[1, 0, 0, 0, 0, 0, 0, 0]);
-        assert_eq!(offset, 0);
-        let offset = pad_and_extend::<i32>(&mut data, &[2, 0, 0, 0]);
-        assert_eq!(offset, 8);
-        let offset = pad_and_extend::<i64>(&mut data, &[3, 0, 0, 0, 0, 0, 0, 0]);
-        assert_eq!(offset, 16);
-        assert_eq!(
-            data,
-            &[1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0]
-        );
+    fn function_signature(&self, index: FunctionIndex) -> Option<VMSharedSignatureIndex> {
+        let imports = self.import_counts.functions as usize;
+        if index.index() < imports {
+            self.imports
+                .iter()
+                .filter_map(|im| {
+                    if let VMImportType::Function(idx) = im.ty {
+                        Some(idx)
+                    } else {
+                        None
+                    }
+                })
+                .nth(index.index())
+        } else {
+            let index = LocalFunctionIndex::new(index.index() - imports);
+            Some(self.functions[index].signature)
+        }
     }
```

Comment:
Ideally this function could go away. I believe it is currently only used for tests, but its `O(n)` and if somebody used it for implementation of something else, it would potentially lead to `O(n^2)`.